common_types/src/row/mod.rs

```diff
@@ -421,6 +394,88 @@ impl<'a> Iterator for IterCol<'a> {
     }
 }
 
+/// Build the [`RowGroup`] from the columns.
+pub struct RowGroupBuilderFromColumn {
+    schema: Schema,
+    cols: HashMap<ColumnId, Vec<Datum>>,
+}
+
+impl RowGroupBuilderFromColumn {
+    pub fn with_capacity(schema: Schema, num_cols: usize) -> Self {
+        Self {
+            schema,
+            cols: HashMap::with_capacity(num_cols),
+        }
+    }
+
+    /// The newly-added column should have the same elements as the
+    /// previously-added column's.
+    pub fn try_add_column(&mut self, col_id: ColumnId, col: Vec<Datum>) -> Result<()> {
+        if let Some(num_rows) = self.num_rows() {
+            ensure!(
+                num_rows == col.len(),
+                InvalidRowNum {
+                    expect: num_rows,
+                    given: col.len(),
+                }
+            );
+        }
+
+        let old = self.cols.insert(col_id, col);
+        ensure!(old.is_none(), DuplicateColumnId { column_id: col_id });
+
+        Ok(())
+    }
+
+    pub fn build(mut self) -> RowGroup {
+        let num_rows = self.num_rows().unwrap_or(0);
+        let num_cols = self.schema.num_columns();
+        let mut rows = Vec::with_capacity(num_rows);
+
+        // Pre-allocate the memory for column data in every row.
+        for _ in 0..num_rows {
+            let row = Vec::with_capacity(num_cols);
+            rows.push(row);
+        }
+
+        let mut add_column_to_row = |row_idx: usize, datum: Datum| {
+            rows[row_idx].push(datum);
+        };
+
+        for col_schema in self.schema.columns() {
+            let col_id = col_schema.id;
+            let datums = self.cols.remove(&col_id);
+
+            match datums {
+                Some(v) => {
+                    for (row_idx, datum) in v.into_iter().enumerate() {
+                        add_column_to_row(row_idx, datum);
+                    }
+                }
+                None => {
+                    for row_idx in 0..num_rows {
+                        add_column_to_row(row_idx, Datum::Null);
+                    }
+                }
+            }
+        }
+
+        let rows = rows.into_iter().map(Row::from_datums).collect::<Vec<_>>();
+
+        RowGroup {
+            schema: self.schema,
+            rows,
+            min_timestamp: Timestamp::now(),
```

Comment:
Seems unnecessary to call `now`, and just use a fixed default?

Reply:
Oh. Maybe the correct timestamp should be computed here.

Reply:
> Oh. Maybe the correct timestamp should be computed here.

It seems time range is unnecessary actually? Maybe default value is enough?

Reply:
Maybe we can't make such an assumption.